### PR TITLE
Feature/rm tests

### PR DESCRIPTION
**Pull request by iparask**
@andre-merzky what do you think? I'm trying to finish the base class of `lm`

### CODE REVIEW

File: src/radical/pilot/states.py [hunk 1]
```diff
@@ -41,13 +41,14 @@
         CANCELED               :  5}
 _pilot_state_inv   = {v: k for k, v in _pilot_state_values.iteritems()}
 _pilot_state_inv_full = dict()
-for s,v in _pilot_state_values.iteritems():
+for st,v in _pilot_state_values.iteritems():
     if v in _pilot_state_inv_full:
```
Comment (andre-merzky) [inline src/radical/pilot/states.py]: why the change from s to st?

File: tests/test_states.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+def test_pilot_state_value():
+    from radical.pilot.states import _pilot_state_value
+    assert(_pilot_state_value('NEW')                    == 0)
+    assert(_pilot_state_value('PMGR_LAUNCHING_PENDING') == 1)
+    assert(_pilot_state_value('PMGR_LAUNCHING')         == 2)
+    assert(_pilot_state_value('PMGR_ACTIVE_PENDING')    == 3)
+    assert(_pilot_state_value('PMGR_ACTIVE')            == 4)
+    assert(_pilot_state_value('DONE')                   == 5)
+    assert(_pilot_state_value('FAILED')                 == 5)
+    assert(_pilot_state_value('CANCELED')               == 5)
```
Comment (andre-merzky) [inline tests/test_states.py]: To be honest, I don't find these tests particularly useful.  The numbers we assign are not really special, so these tests basically assert that `a=b` results in `a=b` - duh ;-)  Not that I have an educated opinion what types of tests would make sense in this context, just saying...

Comment (iparask) [reply tests/test_states.py]: Yeah, but this is the first file in issue #1881. And apparently it is the easiest. I agree with you it is a bit weird to test files that create constants only.

File: src/radical/pilot/states.py (hunk 1 above)
Comment (iparask) [reply src/radical/pilot/states.py]: The linter was complaining because `s` is used also as a parameter to a function.

Comment (andre-merzky) [reply src/radical/pilot/states.py]: Why the hell is it complaining about that?
```
   list.append(s)
   list[len(list)] = s
```
would be equivalent, but it would not complain about the latter?  That does not make sense...  Lets not follow the linter blindly, please...

File: tests/test_states.py (hunk 2 above)
Comment (andre-merzky) [reply tests/test_states.py]: There are a couple of methods there which are supposed to ensure correct state progression.  Those are rather central to how we define and ensure states.  The `collapse` is tested, we may want to add test to the `progress` versions, too?

File: tests/test_states.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+
+# ------------------------------------------------------------------------------
+#
+def test_pilot_state_collapse():
+    from radical.pilot.states import _pilot_state_collapse,_pilot_state_values
+
+    for key in _pilot_state_values.iterkeys():
+        assert(_pilot_state_collapse([key]) == key)
+# ------------------------------------------------------------------------------
```
Comment (andre-merzky) [inline tests/test_states.py]: Looking over this again, I don't think this test covers the right cases.  The method is supposed to select the most current state out of a list of states.  Assume a unit gets state updates out of order (as it happens) - what is our best guess in what the state the unit is in?

```
[NEW, UMGR_STAGING_INPUT, FAILED, UMGR_STAGING_INPUT_PENDING] -> FAILED
[NEW, EXECUTING, UMGR_STAGING_INPUT, UMGR_STAGING_INPUT_PENDING] -> EXECUTING
```
etc.

Comment (iparask) [reply tests/test_states.py]: Aha! Thank you! I will update the test with a few test cases then that have states out of order and provide the last one!

File: tests/test_states.py (hunk 2 above)
Comment (iparask) [reply tests/test_states.py]: Okay!

File: src/radical/pilot/states.py (hunk 1 above)
Comment (iparask) [reply src/radical/pilot/states.py]: The issue was this:
```
for s,v in _pilot_state_values.iteritems():
    if v in _pilot_state_inv_full:
        if not isinstance(_pilot_state_inv_full[v], list):
            _pilot_state_inv_full[v] = [_pilot_state_inv_full[v]]
        _pilot_state_inv_full[v].append(s)
    else:
        _pilot_state_inv_full[v] = s


def _pilot_state_value(s):
    return _pilot_state_values[s]
```
And this was the linter's message:
`W: 53,23: Redefining name 's' from outer scope (line 44) (redefined-outer-name)`

We can take it out if you think this is too much.

File: tests/test_states.py (hunk 3 above)
Comment (andre-merzky) [reply tests/test_states.py]: Thank you, @iparask !

File: src/radical/pilot/states.py (hunk 1 above)
Comment (andre-merzky) [reply src/radical/pilot/states.py]: yes, thank you.

File: tests/test_lm/test_unit/test_rsh.py [hunk 4]
```diff
@@ -62,6 +62,7 @@ def test_construct_command(mocked_init,
     component.launch_command = 'rsh'
 
     for unit, result in test_cases:
+        print unit['uid']
         if result == "ValueError":
```
Comment (andre-merzky) [inline tests/test_lm/test_unit/test_rsh.py]: This can probably go?

File: tests/test_states.py (hunk 2 above)
Comment (andre-merzky) [reply tests/test_states.py]: Should this still be addressed or  will that go into a different PR?  If the latter, please open a ticket to keep track of it.

File: tests/test_states.py (hunk 3 above)
Comment (andre-merzky) [reply tests/test_states.py]: Should this still be addressed or  will that go into a different PR?  If the latter, please open a ticket to keep track of it.

**Comment by andre-merzky** [comment]
@iparask : ping

File: tests/test_lm/test_unit/test_rsh.py (hunk 4 above)
Comment (iparask) [reply tests/test_lm/test_unit/test_rsh.py]: Yes!

File: tests/test_states.py (hunk 3 above)
Comment (iparask) [reply tests/test_states.py]: Fixing it!

**Comment by iparask** [comment]
I think it is ready. Please check once more.

**Comment by iparask** [comment]
Hello @andre-merzky, would you like to have another look?

File: tests/test_cases/unit.000002.json [hunk 5]
```diff
@@ -28,7 +28,8 @@
         "lm": {
             "spark"  : ["/spark-submit    test.py", null],
             "ibrun"  : "RuntimeError",
-            "orte"   : "RuntimeError"
+            "orte"   : "RuntimeError",
```
Comment (mturilli) [inline tests/test_cases/unit.000002.json]: Should we discontinue ORTE altogether?

**Comment by andre-merzky** [comment]
Thanks for all the work!
